Shawn (HyungJoon) Yoon 

0A: Yea, I followed the honor code on this problem set.  

0B: 10hrs 

In [132]:
import pandas as pd
from linearmodels import OLS
import numpy as np

Before starting on this problem set, it may be useful to read through the following description
of the formula language used by `linearmodels`: 

[https://patsy.readthedocs.io/en/v0.1.0/formulas.html#the-formula-language](https://patsy.readthedocs.io/en/v0.1.0/formulas.html#the-formula-language)

In [133]:
# Import data
df_data = pd.read_csv("StaplesData.csv", index_col=0)

# Set parameters
c = 10 # Marginal cost

df_data.head()

Purchase  Price  NotNearCompetingStore  WealthyZipCode  Weekday  \
1         1     13                      0               0        1   
2         1     13                      1               0        1   
3         1     13                      0               0        1   
4         1     13                      0               0        1   
5         0     15                      0               0        1   

   ExistingCustomer  
1                 0  
2                 1  
3                 0  
4                 0  
5                 0

1A: 

$\pi$ = Pr(Purchase) * ($P_i$ - c)

$\pi$ = ($\beta_0$ + $\beta_1$$X_i$ + $\beta_2$$P_i$ + $\beta_3$$X_i$$P_i$) * ($P_i$ - c)

FOC

d$\pi$/d$p$ = $\beta_0$ + $\beta_1$$X_i$ + 2$\beta_2$$P_i$ + 2$\beta_3$$X_i$$P_i$ - $\beta_2$c - $\beta_3$$X_i$c = 0 

Therefore, 

$P_i^*$ = ($\beta_2$c + $\beta_3$$X_i$c - $\beta_0$ - $\beta_1$$X_i$) / (2$\beta_2$ + 2$\beta_3$$X_i$)

In [134]:
# 1B 
covariates = ['NotNearCompetingStore', 'WealthyZipCode', 'Weekday', 'ExistingCustomer']

regression_formula =  'Purchase ~ 1 + Price * ({0})'.format(' + '.join(covariates))
print('regression formula', regression_formula, '\n')

model = OLS.from_formula(regression_formula, data=df_data)
    # regression_formula,
    # something else


res = model.fit()

print(res.summary)

# # store the coefficients as vectors for use later.
beta_0 = res.params.Intercept
beta_1 = res.params[covariates]
beta_2 = res.params.Price
beta_3 = res.params[['Price:' + cov for cov in covariates]]

regression formula Purchase ~ 1 + Price * (NotNearCompetingStore + WealthyZipCode + Weekday + ExistingCustomer) 

                            OLS Estimation Summary                            
Dep. Variable:               Purchase   R-squared:                      0.3230
Estimator:                        OLS   Adj. R-squared:                 0.3228
No. Observations:               40000   F-statistic:                 6.511e+04
Date:                Mon, Oct 07 2019   P-value (F-stat)                0.0000
Time:                        11:15:20   Distribution:                  chi2(9)
Cov. Estimator:                robust                                         
                                                                              
                                      Parameter Estimates                                      
                             Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
--------------------------------------------------------------

#### Coefficient Estimates whose p-value < 10%

$\beta_0$ (intercept) = 2.5703

$\beta_1$ (NotNear) = -1.1598

$\beta_1$ (WealthyZipCode) = -0.1545

$\beta_1$ (Weekday) = 0.1239 

$\beta_2$ (price) = -0.1586

$\beta_3$ (interaction, Price:NotNearCompetingStore) = 0.0596

$\beta_3$ (interaction, Price:WealthyZipCode) = 0.0468

In [135]:
# 1C
# Not being near a competing store
print( 1- (beta_2 + beta_3['Price:NotNearCompetingStore']) / beta_2 ) # reduces by X percent

print(beta_2) # from X percentage points
print(beta_2 + beta_3['Price:NotNearCompetingStore']) # to X percentage points

0.37556495351766905
-0.15857552452319368
-0.09902011502660046


$\beta_2$ = -0.158 (steeper) (near from competing store)

$\beta_2 + \beta_3$ = -0.099 (less steep) (further competing store)

Steepness reduces by 37.55% from $\beta_2$ to $\beta_2 + \beta_3$.



In [136]:
beta_3['Price:NotNearCompetingStore']

0.059555409496593215

$\beta_3$ (Price:NotNearCompetingStore) is +0.059. This means that if a customer is not near a competitor store, the interaction term (Price:NotNearCompetingStore) turns on, and 0.059 gets added to $\beta_2$, making the slope less steeper. So the price responsiveness is smaller than those who are near a competitor store.   

Conditional on other X variables, being not near competing store (further from competing store) reduces price responsiveness by 37.55% points per dollar of price increase.

The signs and magnitudes are consistent with what I expected. $\beta_3$ (Price:NotNearCompetingStore) is +0.059. So, if you add this to $\beta_2$, -0.158, then ($\beta_2$+$\beta_3$) gets bigger. In other words, the slope becomes less steeper. If it is less steeper, then it means that the customer will respond less sensitively to the price change than those who are near the competitor store. 

In [137]:
# Wealthy zip code
print( 1- (beta_2 + beta_3['Price:WealthyZipCode']) / beta_2 ) # reduces by X percent

print(beta_2) # from X percentage points
print(beta_2 + beta_3['Price:WealthyZipCode']) # to X percentage points

0.295430557098287
-0.15857552452319368
-0.11172746897115349


$\beta_2$ = -0.158 (steeper) (poor zip code)

$\beta_2 + \beta_3$ = -0.111 (less steep) (wealthy zip code)

Steepness reduces by 29.54% from $\beta_2$ to $\beta_2 + \beta_3$.

In [138]:
beta_3['Price:WealthyZipCode']

0.04684805555204019

$\beta_3$ (Price:WealthyZipCode) is +0.046. This means that if a customer has a wealthy zipcode, the interaction term (Price:NotNearCompetingStore) turns on, and 0.046 gets added to $\beta_2$, making the slope less steeper. So the price responsiveness is smaller than those who does not have wealthy zip code.   

Conditional on other X variables, being being in a wealthy zip code reduces price responsiveness by 29.54% points per dollar of price increase.

The signs and magnitudes are consistent with what I expected. $\beta_3$ (Price:WealthyZipCode) is +0.046. So, if you add this to $\beta_2$, -0.158, then ($\beta_2$+$\beta_3$) gets bigger. In other words, the slope becomes less steeper. If it is less steeper, then it means that the customer will respond less sensitively to the price change than those who are not in a wealthy zip code. 

In [158]:
#Purchase probability under original price (I am just doing this to check what the profit is if the firm just operated at the given price in the data)
df_data['purchase_proba_original_price'] = beta_0 + np.dot(df_data[covariates], beta_1) + (beta_2 + np.dot(df_data[covariates], beta_3)) * df_data['Price']
df_data['profit_under_original_price'] = (df_data['Price'] - c)*df_data['purchase_proba_original_price']

In [159]:
# 1D: Profit maximizing prices
df_data['monopoly_price'] = c / 2 - (beta_0 + np.dot(df_data[covariates], beta_1)) / (
    2 * (beta_2 + np.dot(df_data[covariates], beta_3)))
df_data.monopoly_price.describe()

count    40000.000000
mean        13.520190
std          1.276742
min         12.122566
25%         13.104497
50%         13.230974
75%         13.455927
max         17.598529
Name: monopoly_price, dtype: float64

In [160]:
# 1E: Expected profit per consumer

#compute purchase proba
df_data['purchase_proba'] = beta_0 + np.dot(df_data[covariates], beta_1) + (beta_2 + np.dot(df_data[covariates], beta_3)) * df_data['monopoly_price']

# something

#Insert the purchase probabilities and the prices into the profit function
df_data['profit'] = df_data['purchase_proba'] * (df_data['monopoly_price'] - c)

# something

print(df_data[['purchase_proba', 'profit']].describe())

       purchase_proba        profit
count    40000.000000  40000.000000
mean         0.472592      1.766739
std          0.142677      1.008556
min          0.210177      0.446114
25%          0.367183      1.528335
50%          0.528805      1.708555
75%          0.528805      1.976572
max          0.728943      4.478494


#### 1F 

$\pi$ = (P - c) * Pr(Purchase) 

$\pi$ = (P - c) * ($\alpha_0$ + $\alpha_2$P)

FOC 

d$\pi$/dP = $\alpha_0$ + 2$\alpha_2$P - $\alpha_2$c = 0

## $p^u$ = c/2 - $\alpha_0$/(2$\alpha_2$)

In [161]:
# 1G: Estimate average demand function

model = OLS.from_formula('Purchase ~ 1 + Price', data=df_data) # something
res = model.fit()

# store the coefficients for use later.
beta_0h = res.params.Intercept
beta_2h = res.params.Price

In [162]:
beta_0h

2.298616586630814

In [163]:
beta_2h

-0.13612331503713904

1H.  

$\alpha_2$ is -0.136. This means that when the firm increases the uniform price by one dollar, the probability of purchase decreases by 0.136.

In [164]:
# 1I: Profit-maximizing uniform price 
uniform_monopoly_price = c / 2 - beta_0h / (2 * beta_2h)
df_data['uniform_price'] = uniform_monopoly_price
print(uniform_monopoly_price) # This just displays the price

13.443140640542268


In [165]:
df_data.head()

Purchase  Price  NotNearCompetingStore  WealthyZipCode  Weekday  \
1         1     13                      0               0        1   
2         1     13                      1               0        1   
3         1     13                      0               0        1   
4         1     13                      0               0        1   
5         0     15                      0               0        1   

   ExistingCustomer  purchase_proba_original_price  \
1                 0                       0.566608   
2                 1                       0.261816   
3                 0                       0.566608   
4                 0                       0.566608   
5                 0                       0.239273   

   profit_under_original_price  monopoly_price  purchase_proba    profit  \
1                     1.699824       13.230974        0.528805  1.708555   
2                     0.785447       12.735689        0.289816  0.792848   
3                     1.699824       13.230974        0.528805  1.708555   
4                     1.699824       13.230974        0.528805  1.708555   
5                     1.196366       13.230974        0.528805  1.708555   

   uniform_price  purchase_proba_unif_price  profit_unif_price  
1      13.443141                    0.49408           1.701188  
2      13.443141                    0.21487           0.739827  
3      13.443141                    0.49408           1.701188  
4      13.443141                    0.49408           1.701188  
5      13.443141                    0.49408           1.701188

In [166]:
# 1J: Expected profit under uniform price
# Here you can use very similar code to 1E, except plug in P_star instead of P_i.
df_data['purchase_proba_unif_price'] = beta_0 + np.dot(df_data[covariates], beta_1) + \
    (beta_2 + np.dot(df_data[covariates], beta_3)) * uniform_monopoly_price


#Insert the purchase probabilities and the prices into the profit function
df_data['profit_unif_price'] = df_data['purchase_proba_unif_price']*(uniform_monopoly_price - c)# something

print(df_data[['purchase_proba_unif_price', 'profit_unif_price']].describe())

       purchase_proba_unif_price  profit_unif_price
count               40000.000000       40000.000000
mean                    0.469590           1.616865
std                     0.242354           0.834458
min                     0.079413           0.273431
25%                     0.438597           1.510150
50%                     0.494080           1.701188
75%                     0.554729           1.910011
max                     1.049369           3.613125


In [167]:
df_data.head()

Purchase  Price  NotNearCompetingStore  WealthyZipCode  Weekday  \
1         1     13                      0               0        1   
2         1     13                      1               0        1   
3         1     13                      0               0        1   
4         1     13                      0               0        1   
5         0     15                      0               0        1   

   ExistingCustomer  purchase_proba_original_price  \
1                 0                       0.566608   
2                 1                       0.261816   
3                 0                       0.566608   
4                 0                       0.566608   
5                 0                       0.239273   

   profit_under_original_price  monopoly_price  purchase_proba    profit  \
1                     1.699824       13.230974        0.528805  1.708555   
2                     0.785447       12.735689        0.289816  0.792848   
3                     1.699824       13.230974        0.528805  1.708555   
4                     1.699824       13.230974        0.528805  1.708555   
5                     1.196366       13.230974        0.528805  1.708555   

   uniform_price  purchase_proba_unif_price  profit_unif_price  
1      13.443141                    0.49408           1.701188  
2      13.443141                    0.21487           0.739827  
3      13.443141                    0.49408           1.701188  
4      13.443141                    0.49408           1.701188  
5      13.443141                    0.49408           1.701188

In [168]:
# 1K: Profit loss from uniform pricing

# something

In [169]:
df_data[['profit', 'profit_unif_price']].head(5)

profit  profit_unif_price
1  1.708555           1.701188
2  0.792848           0.739827
3  1.708555           1.701188
4  1.708555           1.701188
5  1.708555           1.701188

In [170]:
print("Total Profit at price discrimination is", df_data['profit'].sum())
print("Total Profit at uniform price is", df_data['profit_unif_price'].sum())
print("")
print("Therefore, the expected profit loss from uniform pricing instead of price discrimination is", df_data['profit_unif_price'].sum() - df_data['profit'].sum())

Total Profit at price discrimination is 70669.56912811569
Total Profit at uniform price is 64674.60631774961

Therefore, the expected profit loss from uniform pricing instead of price discrimination is -5994.96281036608


In [171]:
print("If the firm operated at the price given in the data, they are going to make only this,", df_data['profit_under_original_price'].sum(), "which is even less than the profit-max uniform price.")

If the firm operated at the price given in the data, they are going to make only this, 57399.000000012464 which is even less than the profit-max uniform price.


1L: What data other than the variables in StaplesData.csv might better predict purchase probability and
thus be useful for price discrimination?



1. Do the consumer has a kid who goes to an elementary school? (yes/no)?

2. Do the consumer go to college? (yes/no)?  

1M: Concisely discuss two or three caveats to this analysis that you (as an economist at Staples) would
want your manager to understand.

1. There are certainly more than  4 variables. But for the given data, we don't oberseve anything more. 

2. Just because we have 4 different variables, it doesn't mean that we are going to have 4 different prices. There are interaction terms, so it means there are going to be a lot more than 4 different prices. 